certbot-ci/certbot_integration_tests/certbot_tests/test_main.py

```diff
@@ -92,9 +90,154 @@ def test_renew_with_hook_scripts(context):
 
     assert_cert_count_for_lineage(context.config_dir, certname, 1)
 
+    misc.generate_test_file_hooks(context.config_dir, context.hook_probe)
+    context.certbot(['renew'])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 2)
+    assert_hook_execution(context.hook_probe, 'deploy')
+
+
+def test_renew_files_propagate_permissions(context):
+    """Test proper certificate renewal with custom permissions propagated on private key."""
+    certname = context.get_domain('renew')
+    context.certbot(['-d', certname])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 1)
+
+    os.chmod(join(context.config_dir, 'archive/{0}/privkey1.pem'.format(certname)), 0o444)
+    context.certbot(['renew'])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 2)
+    assert_equals_permissions(
+        join(context.config_dir, 'archive/{0}/privkey1.pem'.format(certname)),
+        join(context.config_dir, 'archive/{0}/privkey2.pem'.format(certname)), 0o074)
+
+
+def test_graceful_renew_it_is_not_time(context):
+    """Test graceful renew is not done when it is not due time."""
+    certname = context.get_domain('renew')
+    context.certbot(['-d', certname])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 1)
+
+    context.certbot_no_force_renew([
+        'renew', '--deploy-hook', 'echo deploy >> "{0}"'.format(context.hook_probe)])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 1)
+    with pytest.raises(AssertionError):
+        assert_hook_execution(context.hook_probe, 'deploy')
+
+
+def test_graceful_renew_it_is_time(context):
+    """Test graceful renew is done when it is due time."""
+    certname = context.get_domain('renew')
+    context.certbot(['-d', certname])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 1)
+
+    with open(join(context.config_dir, 'renewal/{0}.conf'.format(certname)), 'r') as file:
+        lines = file.readlines()
+    lines.insert(4, 'renew_before_expiry = 100 years{0}'.format(os.linesep))
+    with open(join(context.config_dir, 'renewal/{0}.conf'.format(certname)), 'w') as file:
+        file.writelines(lines)
+
+    context.certbot_no_force_renew([
+        'renew', '--deploy-hook', 'echo deploy >> "{0}"'.format(context.hook_probe)])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 2)
+    assert_hook_execution(context.hook_probe, 'deploy')
+
+
+def test_renew_with_changed_private_key_complexity(context):
+    """Test proper renew with updated private key complexity."""
+    certname = context.get_domain('renew')
+    context.certbot(['-d', certname, '--rsa-key-size', '4096'])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 1)
```

Comment:
nit: Can we also check the size of key1 here?

Reply:
I added the assertion.